packages/trimerge-sync/src/lib/LeaderManager.ts

```diff
@@ -0,0 +1,202 @@
+import { LeaderEvent } from '../types';
+
+type Timeout = ReturnType<typeof setTimeout>;
+type Interval = ReturnType<typeof setInterval>;
+
+export type LeaderSettings = Readonly<{
+  electionTimeoutMs: number;
+  heartbeatMs: number;
+  heartbeatTimeoutMs: number;
+}>;
+
+const DEFAULT_SETTINGS: LeaderSettings = {
+  electionTimeoutMs: 200,
+  heartbeatMs: 1_000,
+  heartbeatTimeoutMs: 2_500,
+};
+
+/**
+ * This class takes a best-effort approach to picking a single leader amongst
+ * multiple clients assuming a shared broadcast channel.
+ *
+ * Whenever a client comes online, it requests an election.
+ *
+ * 1. If there is already a leader, it will announce itself, cancelling the election.
+ * 2. Otherwise, everyone announces themselves as available.
+ *
+ * After an election timeout, the client with the alphanumerically smallest
+ * clientId announces itself as the leader to everyone.
+ *
+ * If a leader receives an announcement from another leader, a re-election will
+ * occur, and all but one of them will be demoted back to non-leader.
+ */
+export class LeaderManager {
+  private closed = false;
+  private currentLeaderId?: string = undefined;
+  private isLeader: boolean = false;
+  private potentialLeaders = new Set<string>();
+  private electionTimeout?: Timeout = undefined;
+  private leaderHeartbeat?: Interval = undefined;
+  private heartbeatTimeout?: Timeout = undefined;
+
+  constructor(
+    /**
+     * A unique identifier for each client. This can simply be a uuid
+     */
+    private readonly clientId: string,
+    /**
+     * A callback for when this client gains or loses leadership
+     */
+    private readonly onLeaderChange: (isLeader: boolean) => void,
+    /**
+     * A callback for when this class needs to broadcast messages to other clients
+     */
+    private readonly broadcastEvent: (event: LeaderEvent) => void,
+    private readonly settings: LeaderSettings = DEFAULT_SETTINGS,
+  ) {
+    this.elect();
+  }
+
+  private elect() {
+    if (this.electionTimeout) {
+      return;
+    }
+    this.potentialLeaders.clear();
+    this.potentialLeaders.add(this.clientId);
+    this.broadcastEvent({
+      type: 'leader',
+      action: 'request',
+      clientId: this.clientId,
+    });
+    this.electionTimeout = setTimeout(
+      () => this.finishElection(),
+      this.settings.electionTimeoutMs,
+    );
+  }
+
+  private cancelElection() {
+    if (this.electionTimeout) {
+      clearTimeout(this.electionTimeout);
+      this.electionTimeout = undefined;
+    }
+  }
+  private finishElection() {
+    this.electionTimeout = undefined;
+    const [proposedLeader] = Array.from(this.potentialLeaders).sort();
```

Comment:
nit: This will be O(n log n) but could be O(n) since you only need to find the first one.

Reply:
Another nit: TypeScript will infer `proposedLeader` as a string here, when it's really a `string | undefined`.

Reply:
okay, fixing both